### PR TITLE
net: dhcpv4: Implement DHCPv4 server

### PR DESCRIPTION
**Pull request by rlubos**
A simple, socket-based DHCPv4 server implementation based on RFC2131.

### CODE REVIEW

File: include/zephyr/net/net_if.h [hunk 1]
```diff
@@ -2279,6 +2279,15 @@ struct in_addr *net_if_ipv4_get_ll(struct net_if *iface,
...
 					    enum net_addr_state addr_state);
 
+/**
+ * @brief Get IPv4 netmask of an interface.
+ *
+ * @param iface Interface to use.
+ *
+ * @return A pointer to the netmask set on the interface, NULL if not found.
+ */
+struct in_addr *net_if_ipv4_get_netmask(struct net_if *iface);
```
Comment (jukkar) [inline include/zephyr/net/net_if.h:2289]: Just wondering here that should we just return the in_addr which is basically a uint32_t type variable. Will the caller of this function need a pointer the original value?

File: subsys/net/lib/shell/dhcpv4.c [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+	if (lease->state == DHCPV4_SERVER_ADDR_DECLINED) {
+		snprintk(expiry_str, sizeof(expiry_str) - 1, "infinite");
+	} else {
+		snprintk(expiry_str, sizeof(expiry_str) - 1, "%u",
+			 timepoint_to_s(lease->expiry));
+	}
+
+	PR("%2d. %15s\t%9s\t%12s\n", *count, net_sprint_ipv4_addr(&lease->addr),
+	   dhcpv4_addr_state_to_str(lease->state), expiry_str);
+
```
Comment (jukkar) [inline subsys/net/lib/shell/dhcpv4.c:143]: Extra empty line can be removed.

File: subsys/net/lib/shell/dhcpv4.c [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	if (ret == -ENOENT) {
+		PR_WARNING("DHCPv4 server is not running on interface %d\n", idx);
+	} else if (ret < 0) {
+		PR_ERROR("DHCPv4 server failed to stop on interface %d, error %d\n",
+			 idx, -ret);
+	} else {
+		PR("DHCPv4 server stopped on interface %d\n", idx);
+	}
+#else /* CONFIG_NET_DHCPV4_SERVER */
+	PR_INFO("Set CONFIG_NET_DHCPV4_SERVER to enable DHCPv4 server support.\n");
```
Comment (jukkar) [inline subsys/net/lib/shell/dhcpv4.c:80]: I suggest we follow the same pattern as in other net-shell files so save some memory.
So
```
	PR_INFO("Set %s to enable %s support.\n",
		"CONFIG_NET_DHCPV4_SERVER",
		"DHCPv4 server");
```
Same comment to the start command.

File: subsys/net/lib/shell/dhcpv4.c [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	user_data.user_data = &count;
+
+	ret = net_dhcpv4_server_foreach_lease(iface, dhcpv4_lease_cb, &user_data);
+	if (ret == -ENOENT) {
+		PR_WARNING("DHCPv4 server is not running on interface %d\n", idx);
+	} else if (count == 0) {
+		PR("DHCPv4 server - no addresses assigned on interface %d\n", idx);
+	}
+#else /* CONFIG_NET_DHCPV4_SERVER */
+	PR_INFO("Set CONFIG_NET_DHCPV4_SERVER to enable DHCPv4 server support.\n");
```
Comment (jukkar) [inline subsys/net/lib/shell/dhcpv4.c:176]: ditto

File: subsys/net/lib/shell/dhcpv4.c [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+);
+
+SHELL_STATIC_SUBCMD_SET_CREATE(net_cmd_dhcpv4,
+	SHELL_CMD(server, &net_cmd_dhcpv4_server,
+		  "DHCPv4 server service management.",
+		  NULL),
+	SHELL_SUBCMD_SET_END
+);
+
+SHELL_SUBCMD_ADD((net), dhcpv4, &net_cmd_dhcpv4, "Manage DHPCv4 services.",
```
Comment (jukkar) [inline subsys/net/lib/shell/dhcpv4.c:205]: Should we have also some commands for the DHCPv4 client? I recently noticed that if I enable DHCPv4 but disable config module, then I need to add a call to `net_dhcpv4_start/stop` to some file. For debugging purposes it would be useful to support the client side in this shell. Naturally this could be in another PR.

**Review by jukkar — COMMENTED** [review]
Some preliminary comments, will continue the review later.

File: subsys/net/l2/ethernet/arp.h [hunk 6]
```diff
@@ -83,7 +86,10 @@ void net_arp_init(void);
 #define net_arp_foreach(...) 0
 #define net_arp_init(...)
 #define net_arp_clear_pending(...) 0
+#define net_arp_update(...)
 
 #endif /* CONFIG_NET_ARP */
 
+
+
```
Comment (pdgendt) [inline subsys/net/l2/ethernet/arp.h:94]: Nit: remove newlines

File: subsys/net/lib/dhcp_server/dhcpv4_server.c [hunk 7]
```diff
@@ -0,0 +1,1423 @@
...
+
+	opt = dhcpv4_find_option(data, datalen, DHCPV4_OPTIONS_MSG_TYPE);
+	if (opt == NULL) {
+		return -ENOENT;
+	}
+
+	optlen = *(opt + 1);
+	if (optlen != 1) {
+		return -EINVAL;
+	}
```
Comment (pdgendt) [inline subsys/net/lib/dhcp_server/dhcpv4_server.c:146]: A minor optimization could be to pass an `optlen` pointer to `dhcpv4_find_option`, so you don't have to repeat that each time in these specialized functions.

**Review by pdgendt — COMMENTED** [review]
I think we need some integration tests for the DHCPv4 server?

File: subsys/net/lib/dhcp_server/dhcpv4_server.c [hunk 8]
```diff
@@ -0,0 +1,1423 @@
...
+
+int net_dhcpv4_server_start(struct net_if *iface, struct in_addr *base_addr)
+{
+	struct sockaddr_in addr = {
+		.sin_family = AF_INET,
+		.sin_addr = INADDR_ANY_INIT,
+		.sin_port = htons(DHCPV4_SERVER_PORT),
+	};
+	struct ifreq ifreq = { 0 };
+	bool start = true;
```
Comment (pdgendt) [inline subsys/net/lib/dhcp_server/dhcpv4_server.c:1207]: Variable doesn't seem to be used?

File: subsys/net/lib/dhcp_server/dhcpv4_server.c [hunk 9]
```diff
@@ -0,0 +1,1423 @@
...
+	}
+
+	return 0;
+}
+
+static int dhcpv4_send_offer(struct dhcpv4_server_ctx *ctx, struct dhcp_msg *msg,
+			     struct in_addr *addr, uint32_t lease_time,
+			     struct dhcpv4_parameter_request_list *params)
+{
+	uint8_t reply[NET_IPV4_MTU];
```
Comment (jukkar) [inline subsys/net/lib/dhcp_server/dhcpv4_server.c:519]: Could we do with smaller buffer, it is now 576 bytes long?
One option could be define the buffer size in kconfig and let user to adjust it.

Comment (rlubos) [reply subsys/net/lib/dhcp_server/dhcpv4_server.c]: I didn't see any specific requirement for the server in the RFC2131 (it's quite chaotic to be honest), but for the client it says:
> The 'options' field is now variable length. A DHCP client must be
   prepared to receive DHCP messages with an 'options' field of at least
   length 312 octets.  This requirement implies that a DHCP client **must
   be prepared to receive a message of up to 576 octets** , the minimum IP
   datagram size an IP host must be prepared to accept
   
Hence my rationale to use IPv4_MTU for the buffer size. In practice, I see in wireshark that Discover/Request sent by th client are 342 bytes long, Offer/Ack sent in response are 304 bytes, so there is a little room for improvement if we find it valuable to optimize a bit. WDYT, is it worth it?

File: subsys/net/lib/CMakeLists.txt [hunk 10]
```diff
@@ -15,6 +15,7 @@ add_subdirectory_ifdef(CONFIG_NET_CAPTURE            capture)
 add_subdirectory_ifdef(CONFIG_NET_ZPERF              zperf)
 add_subdirectory_ifdef(CONFIG_NET_SHELL              shell)
 add_subdirectory_ifdef(CONFIG_NET_TRICKLE            trickle)
+add_subdirectory_ifdef(CONFIG_NET_DHCPV4_SERVER      dhcp_server)
```
[the remark this answers is not in the document]
Comment (rlubos) [reply subsys/net/lib/CMakeLists.txt]: Good suggestion, I've renamed the directory as suggested.

File: subsys/net/lib/dhcp_server/dhcpv4_server.c (hunk 8 above)
Comment (rlubos) [reply subsys/net/lib/dhcp_server/dhcpv4_server.c]: Yes, a leftover from before I've switched to use socket services, thanks.

File: subsys/net/lib/dhcp_server/dhcpv4_server.c (hunk 7 above)
Comment (rlubos) [reply subsys/net/lib/dhcp_server/dhcpv4_server.c]: Good idea, thanks.

File: subsys/net/lib/shell/dhcpv4.c (hunk 3 above)
Comment (rlubos) [reply subsys/net/lib/shell/dhcpv4.c]: Thanks, updated.

File: subsys/net/lib/shell/dhcpv4.c (hunk 5 above)
Comment (rlubos) [reply subsys/net/lib/shell/dhcpv4.c:208]: Yes, I thought it'd be useful to have commands for DHCP client as well, didn't want to mix them in here with the server specific stuff. I can prepare a PR with new shell commands and the source file relocation once we're done with the server.

File: include/zephyr/net/net_if.h (hunk 1 above)
Comment (rlubos) [reply include/zephyr/net/net_if.h]: I don't need a pointer for the DHCP case, maybe indeed it makes more sense to return a value here, changed.

File: subsys/net/lib/dhcp_server/dhcpv4_server.c (hunk 9 above)
Comment (jukkar) [reply subsys/net/lib/dhcp_server/dhcpv4_server.c]: Dunno, and this can be optimized later too. Perhaps we can leave that like this and change it later if needed.

**Comment by rlubos** [comment]
> I think we need some integration tests for the DHCPv4 server?

@pdgendt Definitely, I'll be working on tests now, just opened a PR for early review since it's quite large already.

@pdgendt @jukkar I think I've addressed the first batch of your comments, thanks.

File: subsys/net/lib/dhcpv4/dhcpv4_server.c [hunk 11]
```diff
@@ -0,0 +1,1421 @@
...
+
+		len = *(data + 1);
+
+		if (datalen < len + 2) {
+			break;
+		}
+
+		if (code == opt_code) {
+			/* Found the option. */
+			opt = data;
```
Comment (pdgendt) [inline subsys/net/lib/dhcpv4/dhcpv4_server.c:121]: Nit: Already set `opt` to the correct offset?

```suggestion
			opt = data + 2;
```

**Review by pdgendt — DISMISSED** [review]

File: subsys/net/lib/dhcpv4/dhcpv4_server.c (hunk 11 above)
Comment (rlubos) [reply subsys/net/lib/dhcpv4/dhcpv4_server.c]: Right, done.

**Review by pdgendt — DISMISSED** [review]

File: subsys/net/lib/dhcpv4/dhcpv4_server.c [hunk 12]
```diff
@@ -0,0 +1,1421 @@
...
+						  ARRAY_SIZE(fds), NULL);
+	}
+
+out:
+	k_mutex_unlock(&server_lock);
+
+	return ret;
+}
+
+int net_dhcpv4_server_foreach_lease(struct net_if *iface,
```
Comment (jukkar) [inline subsys/net/lib/dhcpv4/dhcpv4_server.c:1381]: Just wondering should there be a way to go through all the leases over all interfaces, for example user could set the iface to NULL to achieve this?

File: subsys/net/lib/shell/dhcpv4.c [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+static int cmd_net_dhcpv4_server_status(const struct shell *sh, size_t argc, char *argv[])
+{
+#if defined(CONFIG_NET_DHCPV4_SERVER)
+	struct net_shell_user_data user_data;
+	struct net_if *iface = NULL;
+	int idx, ret;
+	int count = 0;
+
+	idx = get_iface_idx(sh, argv[1]);
```
Comment (jukkar) [inline subsys/net/lib/shell/dhcpv4.c:156]: One improvement here could be that if user does not supply interface, the code could call the foreach function with iface NULL and get information about all the interfaces. Could be done in future PR of course.

**Review by jukkar — DISMISSED** [review]

File: subsys/net/lib/shell/dhcpv4.c (hunk 13 above)
Comment (rlubos) [reply subsys/net/lib/shell/dhcpv4.c:156]: That could be a good improvement, I often found myself forgetting the iface index when checking the status. Will do.

File: subsys/net/lib/dhcpv4/dhcpv4_server.c (hunk 12 above)
Comment (rlubos) [reply subsys/net/lib/dhcpv4/dhcpv4_server.c:1394]: The function now accepts NULL iface, and loops over all active server instances.

File: subsys/net/lib/shell/dhcpv4.c (hunk 13 above)
Comment (rlubos) [reply subsys/net/lib/shell/dhcpv4.c]: Done, please re-check.

**Comment by fabiobaltieri** [comment]
Hi, the CI failure has been fixed upstream, if you rebase it should pass.

**Review by jukkar — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]
